pkg/agent/multicluster/mc_route_controller.go

```diff
@@ -0,0 +1,443 @@
+	tunnelPeerIP := getPeerGatewayIP(ciImport.Spec)
+	klog.InfoS("Adding/updating remote Gateway Node flows for Multi-cluster", "gateway", klog.KObj(activeGW),
+		"node", c.nodeConfig.Name, "peer", tunnelPeerIP)
+	allCIDRs := []string{ciImport.Spec.ServiceCIDR}
+	peerConfigs, err := generatePeerConfigs(allCIDRs, tunnelPeerIP)
+	if err != nil {
+		klog.ErrorS(err, "Parse error for serviceCIDR from remote cluster", "clusterinfoimport", ciImport.Name, "gateway", activeGW.Name)
+		return err
+	}
+	if activeGW.Name == c.nodeConfig.Name {
+		klog.V(2).InfoS("Adding/updating flows to remote Gateway Node for Multi-cluster traffic", "clusterinfoimport", ciImport.Name, "cidrs", allCIDRs)
+		localGatewayIP := net.ParseIP(activeGW.GatewayIP)
+		if err := c.ofClient.InstallMulticlusterGatewayFlows(
+			activeGW.Name+ciImport.Name,
+			peerConfigs,
+			tunnelPeerIP,
+			localGatewayIP); err != nil {
+			return fmt.Errorf("failed to install flows to remote Gateway in ClusterInfoImport %s: %v", ciImport.Name, err)
+		}
+	} else {
+		klog.V(2).InfoS("Adding/updating flows to the local active Gateway for Multi-cluster traffic", "clusterinfoimport", ciImport.Name, "cidrs", allCIDRs)
+		tunnelPeerIP := net.ParseIP(activeGW.InternalIP)
+		if err := c.ofClient.InstallMulticlusterNodeFlows(
+			activeGW.Name+ciImport.Name,
+			peerConfigs,
+			tunnelPeerIP); err != nil {
+			return fmt.Errorf("failed to install flows to Gateway %s: %v", activeGW.Name, err)
+		}
+	}
+
+	c.installedCIImports.Add(ciImport)
+	return nil
+}
+
+func (c *MCRouteController) deleteMCNodeFlowsForCIImport(ciImp *mcv1alpha1.ClusterInfoImport, activeGWName string) error {
+	key := activeGWName + ciImp.Name
+	if err := c.ofClient.UninstallMulticlusterFlows(key); err != nil {
+		return fmt.Errorf("failed to uninstall multi-cluster flows to remote Gateway Node %s: %v", ciImp.Name, err)
+	}
+	return nil
+}
+
+func (c *MCRouteController) deleteMCNodeFlowsForGateway(gwName string) error {
+	cachedciImps := c.installedCIImports.List()
+	for _, ciImp := range cachedciImps {
+		ciImport := ciImp.(*mcv1alpha1.ClusterInfoImport)
+		key := gwName + ciImport.Name
+		if err := c.ofClient.UninstallMulticlusterFlows(key); err != nil {
+			return fmt.Errorf("failed to uninstall multi-cluster flows to remote Gateway in ClusterInfoImport %s: %v",
+				ciImport.Name, err)
+		}
+		c.installedCIImports.Delete(ciImport)
+	}
+	return nil
+}
+
+// getActiveGateway compares Gateway's CreationTimestamp to get the active Gateway,
+// The last created Gateway will be the active Gateway.
+func (c *MCRouteController) getActiveGateway() (*mcv1alpha1.Gateway, error) {
+	gws, err := c.gwLister.Gateways(c.namespace).List(labels.Everything())
+	if err != nil {
+		return nil, err
+	}
+	if len(gws) == 0 {
+		return nil, nil
+	}
+	// Comparing Gateway's CreationTimestamp to get the last created Gateway.
+	lastCreatedGW := gws[0]
```

Comment:
If it's possible that a gateway has no gatewayIP or internalIP, it needs to check the fields too.